typegen/src/utils.rs

```diff
@@ -330,6 +384,380 @@ mod tests {
+                    ],
+                    type_def: TypeDef::Composite(TypeDefComposite::new([
+                        Field::new(None, meta_type::<u8>(), Some("A"), Vec::new()),
+                        Field::new(None, meta_type::<u16>(), Some("D"), Vec::new()),
+                        Field::new(None, meta_type::<u32>(), Some("B"), Vec::new()),
+                    ])),
+                    docs: vec![],
+                }
+            }
+        }
+
+        struct D;
+        impl scale_info::TypeInfo for D {
+            type Identity = Self;
+            fn type_info() -> scale_info::Type {
+                scale_info::Type {
+                    path: Path::new("Foo", "my::module"),
+                    type_params: vec![TypeParameter::new("A", Some(meta_type::<u8>()))],
+                    type_def: TypeDef::Composite(TypeDefComposite::new([Field::new(
+                        None,
+                        meta_type::<u8>(),
+                        Some("A"),
+                        Vec::new(),
+                    )])),
+                    docs: vec![],
+                }
+            }
+        }
+        struct E;
+        impl scale_info::TypeInfo for E {
+            type Identity = Self;
+            fn type_info() -> scale_info::Type {
+                scale_info::Type {
+                    path: Path::new("Foo", "my::module"),
+                    type_params: vec![TypeParameter::new("B", Some(meta_type::<u16>()))],
+                    type_def: TypeDef::Composite(TypeDefComposite::new([Field::new(
+                        None,
+                        meta_type::<u16>(),
+                        Some("B"),
+                        Vec::new(),
+                    )])),
+                    docs: vec![],
+                }
+            }
+        }
+
+        let mut registry = scale_info::Registry::new();
+        let id_b = registry.register_type(&meta_type::<B>()).id;
+        let id_bb = registry.register_type(&meta_type::<BB>()).id;
+        let id_a = registry.register_type(&meta_type::<A>()).id;
+        let id_c = registry.register_type(&meta_type::<C>()).id;
+
+        let id_d = registry.register_type(&meta_type::<D>()).id;
+        let id_e = registry.register_type(&meta_type::<E>()).id;
+
+        nested_type!(Y, A, A);
+        nested_type!(W, B, B);
+        nested_type!(Z, C, C);
+
+        let id_y = registry.register_type(&meta_type::<Y>()).id;
+        let id_w = registry.register_type(&meta_type::<W>()).id;
+        let id_z = registry.register_type(&meta_type::<Z>()).id;
+
+        let mut registry = PortableRegistry::from(registry);
+
+        // A != B, different field ordering
+        assert!(!types_equal(id_a, id_b, &registry));
+        assert!(!types_equal(id_a, id_bb, &registry));
+
+        // A == C, different generic param names
+        assert!(types_equal(id_a, id_c, &registry));
+
+        // D == E, different generic param names
```

Comment:
Could we simnplify anything by giving up this constraint? BEcause personally I don't mind if we end up with duplicate types when the generic params are different (because if the generic params are different then, even though it might be unifiable, it's definitely a different type definition :))

Reply:
It's the freebie we get when we use generic indexes, basically no effort was done for it to work, it will just work by default